### PR TITLE
rename DisableTorchFunction to DisableTorchFunctionSubclass

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88219
* __->__ #88218

First half of #87990. This doesn't change any of the behavior and is just a rename

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
This is probably de facto BC breaking, as we never gave a public API for this. Can we design a public API?

**Comment by ezyang** [comment]
Oh I see in the next PR you change the semantics of the old API 🤔

**Comment by ezyang** [comment]
One Q that I would have is as a developer, when should I use the subclass only API, as opposed to all the APIs. In fact, it seems more "reasonable" to be able to disable all modes but NOT the subclass, in cases where you want to stop whatever dynamically scoped black magic is going on, but tensor subclasses should still work the ordinary way

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -506,7 +506,7 @@ def wrap_tensor(self, value: torch.Tensor):
                 )
             # Disable __torch_function__ to prevent cloning of `value` to hit
             # us
-            with torch._C.DisableTorchFunction():
+            with torch._C.DisableTorchFunctionSubclass():
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:509]: I'm not sure but I kind of suspect they also wanted modes disabled too

File: torch/distributed/_shard/partial_tensor.py [hunk 2]
```diff
@@ -236,7 +236,7 @@ def find_process_group(e):
         # Need to disable all dispatch to print args and kwargs appropriately.
```
Comment (ezyang) [inline torch/distributed/_shard/partial_tensor.py:236]: Preexisting condition: this looks questionable

**Review by ezyang — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]

**Comment by samdow** [comment]
> One Q that I would have is as a developer, when should I use the subclass only API, as opposed to all the APIs. In fact, it seems more "reasonable" to be able to disable all modes but NOT the subclass, in cases where you want to stop whatever dynamically scoped black magic is going on, but tensor subclasses should still work the ordinary way

Sure and I can change the dynamo usages after the next PR and see what happens. I think my big worry was that currently this currently works for them and so we would be silently changing their behavior out from under them. With dynamo too, FakeTensor has a torch function mode so I wasn't actually certain if there were places where they wanted only the function mode and no subclasses

**Comment by ezyang** [comment]
I absolutely agree that from a risk reduction perspective this is the right way to structure the change

**Comment by izaitsevfb** [comment]
Hi Samantha! 

Your PyTorch PR imported as [D41211901](https://www.internalfb.com/intern/diff/D41211901/) broke some internal tests:
```
ImportError: cannot import name 'DisableTorchFunction' from 'torch._C' (/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/e84a29bb8ca97004/executorch/exir/tests/__graph__/graph#link-tree/torch/_C.so)
```
due to BC-breaking API change. I have to revert it (with another PR on top of it) to unblock the diff train. Thank you for understanding!

**Pull request closed without merging**